lib/src/util/component_usage.dart

```diff
@@ -23,31 +40,495 @@ class FluentComponentUsage {
+  ///
+  /// See also: other `cascaded`* methods in this class.
+  Iterable<PropAssignment> get cascadedProps => cascadeSections
+      .whereType<AssignmentExpression>()
+      .where((assignment) => assignment.leftHandSide is PropertyAccess)
+      .map((assignment) => PropAssignment(assignment));
+
+  /// The prop reads cascaded on this usage's builder.
+  ///
+  /// See also: other `cascaded`* methods in this class.
+  Iterable<PropRead> get cascadedGetters =>
+      cascadeSections.whereType<PropertyAccess>().map((p) => PropRead(p));
+
+  /// The index prop assignments cascaded on this usage's builder.
+  ///
+  /// See also: other `cascaded`* methods in this class.
+  Iterable<IndexPropAssignment> get cascadedIndexAssignments => cascadeSections
+      .whereType<AssignmentExpression>()
+      .where((assignment) => assignment.leftHandSide is IndexExpression)
+      .map((assignment) => IndexPropAssignment(assignment));
+
+  /// The method calls cascaded on this usage's builder.
+  ///
+  /// See also: other `cascaded`* methods in this class.
+  Iterable<BuilderMethodInvocation> get cascadedMethodInvocations =>
+      cascadeSections
+          .whereType<MethodInvocation>()
+          .map((methodInvocation) => BuilderMethodInvocation(methodInvocation));
 
-  /// The number of child arguments passed into the invocation.
-  int get childArgumentCount => node.argumentList.arguments.length;
+  /// All the cascades on this usage's builder, each wrapped in one of the
+  /// [BuilderMemberAccess] subtypes.
+  ///
+  /// See also: other `cascaded`* methods in this class.
+  List<BuilderMemberAccess> get cascadedMembers {
+    final allHandledMembers = [
+      ...cascadedProps,
+      ...cascadedGetters,
+      ...cascadedIndexAssignments,
+      ...cascadedMethodInvocations,
+    ];
+
+    final allUnhandledMembers = cascadeSections
+        .where((section) => !allHandledMembers.any((m) => m.node == section))
+        .map((section) => OtherBuilderMemberAccess(section))
+        .toList();
+
+    return [
+      ...allHandledMembers,
+      ...allUnhandledMembers,
+    ]
+      // Order them in the order they appear by sorting based on the offset.
+      ..sort((a, b) => a.node.offset.compareTo(b.node.offset));
+  }
+}
+
+/// A child of a component usage.
+///
+/// Will be either an instance of:
+/// - [ExpressionComponentChild]
+/// - [CollectionElementComponentChild] (more of an edge case)
+abstract class ComponentChild {
+  /// The root node of the child.
+  AstNode get node;
+}
+
+/// A child declared as a [CollectionElement] within a single list literal argument
+/// to a component usage.
+///
+/// Example:
+/// ```dart
+/// Dom.div()([
+///   // ExpressionComponentChild
+///   'foo',
+///   // CollectionElementComponentChild
+///   if (condition) 'bar',
+/// ])
+/// ```
+class CollectionElementComponentChild implements ComponentChild {
+  @override
+  final CollectionElement node;
+
+  CollectionElementComponentChild(this.node);
 }
```

Comment:
This is also kinda in relation to `ExpressionComponentChild` what happens in situations like
```
Dom.div()(
'foo',
['bar'],
)
```
what would bar get marked as in this situation? and does `isVariadic` apply?

Reply:
In that case,  you'd have two `ExpressionComponentChild`s, one for `'foo'` and one for `['bar']`